### ISSUE TITLE
Proposal: If the backend returns an alert and results, surface both in the webapp

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#9745]
Currently, if the GQL return value for search results contains both valid results _and_ an alert, the alert is silently suppressed (not shown in the webapp).

The webapp itself has logic to raise alerts and show results, e.g., when raising the [`"Sourcegraph.com searches over the top 10k GitHub repositories..."`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8d35afcc278fbd925d4d4a86913c10a1f0d59af4/-/blob/web/src/marketing/ServerBanner.tsx#L20:14&tab=references) for sourcegraph.com, it still shows results. It just needs to do functionally the same thing if it finds there is an alert in the [GQL response from the backend](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@932ff96a30f9fd5d283d085ab1525c806461dbf9/-/blob/cmd/frontend/graphqlbackend/schema.graphql#L1552-1553). 

Note that other search engines provide this type of things also (e.g., grep.app will say "here are results, note these are partial"), and we could benefit from similar use cases. 

We should preserve the current importance of alerts by keeping the description as

> An alert message that should be displayed _before_ any results.

so that clients continue to understand that alerts should continue to be checked first. The only change is that they _may_ now look at the result set (which we ensure is valid, although perhaps partial), and use it.

~Corresponding to this change, I'd also propose a change to the GQL description to reword~

~> An alert message that should be displayed _before_ any results.~

~to~

~> An alert message that should be displayed _along with_ any results.~

**Comment by slimsag** [comment]
> Currently, if the GQL return value for search results contains both valid results and an alert, the alert is silently suppressed (not shown in the webapp).

I don't believe this is true, because we don't ever return both valid results and an alert right now. It's not possible currently for our backend to do this AFAIK.

I am generally in favor of this proposal, but some questions/concerns first:

1. If we do start returning valid results and an alert, where we otherwise would've returned just valid results, how do we communicate this to existing API consumers? (this would be a backwards-incompatible change)
2. The example of "partial results", and many other cases that I can imagine would have alerts, are already covered by explicit GraphQL fields such as these: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8d35afcc278fbd925d4d4a86913c10a1f0d59af4/-/blob/cmd/frontend/graphqlbackend/schema.graphql#L1530-1551 which for me raises these questions:
    - Do we deprecate `timedout`, `cloning`, etc. and move those into alerts in the long term?
    - Do we make `alert` have a type that can be programmatically handled? And if not, how do we ensure we don't introduce alerts that need to be reasonably consumed in scripts?

**Comment by rvantonder** [comment]
> I don't believe this is true, because we don't ever return both valid results and an alert right now.

I set a `SearchResultsResolver` containing both an alert and results/filematches to confirm this is indeed the case.

> It's not possible currently for our backend to do this AFAIK.

It's certainly possible to set both, but we may not be doing it in any backend functionality because of an implicit belief that it isn't supported or possible, meaning it hasn't been done. I've run into a use case, and found myself wanting this before, but I'll defer on describing exact use cases, since you're in favor.

For 1: 

> how do we communicate this to existing API consumers? 

 I think we should communicate changes to search functionality like we currently do in the CHANGELOG/blog post:  "responses containing alerts _may_ now be accompanied by valid results" (but never invalid results) as well as functionality changes that raise alerts and/or possibly returns result, e.g., "we limit searching over more than 10000 repos will now raise an alert". 

Thinking about it, the phrasing proposal in the description should stay as "_before_", since client should continue to check alerts before doing anything else, as our description currently implies.
I am not sure that this is then a backwards incompatible change, since now clients _may_ make use of the new possibility to check or use results accompanied by an alert, but we can always choose to not send any results, and the client must accept that. Imagine we want to go back on this 'feature': we'd simply 
not return results with the alert. I.e., even if clients decide to make use or rely on this possibility, I don't see how it would be any more backwards incompatible than, say, us deciding to limit results to only 50 repos for diff/commit search, and changing that to 100 at a later stage. This is because the relative importance/precedence of alerts wrt the API would be unchanged: clients currently check alerts before showing any results (as they should according to the description).

For 2: This seems more open-ended and maybe there could be a migration/adaption of existing notifications. It should be considered, but doesn't seem like something that needs to be addressed or answered right for this proposal. Will think about it, but others probably have a reasonable idea here too.

**Comment by slimsag** [comment]
> we may not be doing it in any backend functionality because of an implicit belief that it isn't supported or possible, meaning it hasn't been done.

Yes, this is what I meant -- it has been our policy (up until now) to not do this -- we just don't have any linter or panic etc. that prevents it.

Describing your use case here may give me more confidence that this is a good change; my main concern here can be simplified to the following:

Currently, clients treat the presence of an `alert` as an error ("the search didn't return any results for reason X (described by the alert)"). If that interpretation in existing clients remains valid, then this isn't a substantial backwards-incompatible change. If, however, we for add a single alert that violates this assumption then users could be extremely surprised and many/all existing usages of our search API / `src` CLI may need to be updated. This seems easy for us to accidentally introduce if the webapp does display alerts and results.

Example: a change like "display an alert if a single repository in the search timed out" would work after this change, and would be very nice in the webapp -- but would break all existing clients which assume alert == no results / an error.

I fear the ambiguity here may come back to harm us

**Comment by unknwon** [comment]
I like this proposal from a user's POV, it is better UX.

> Example: a change like "display an alert if a single repository in the search timed out" would work after this change, and would be very nice in the webapp -- but would break all existing clients which assume alert == no results / an error.
> 
> I fear the ambiguity here may come back to harm us

I agree this is a valid concern, to overcome this, could we try the similar approach that GitHub API uses, which sends a special value in `Accept` header, e.g. `application/alert+partial-results` (just made up, but you get the idea).

**Comment by rvantonder** [comment]
> Describing your use case here may give me more confidence that this is a good change

It's really a very practical and familiar thing related to the general idea of search. If you go type `corsca` into Google, you'll see:

<img width="331" alt="Screen Shot 2020-04-14 at 4 51 34 PM" src="https://user-images.githubusercontent.com/888624/79284694-435fa500-7e70-11ea-8231-efbd9e48b000.png">

I want the same thing our search, not necessarily just alert suggestions or 'search instead for...' but also 'I interpreted your query in this way, because this is what you likely meant'. This comes up in particular for and/or queries, around parentheses and disambiguation and UX, and what prompts my proposal. Rather than complaining about a syntax error or ambiguity in operator precedence, we can run the search for what we think it might be, and communicate that along with results (along with suggestions, or not, or whatever). It doesn't need to be this huge alert banner thing, just a small nudge/notification (this is also why I prompted revisiting the alert prompt design).

> Example: a change like "display an alert if a single repository in the search timed out" would work after this change, and would be very nice in the webapp -- but would break all existing clients which assume alert == no results / an error.

I question to what extent

> clients which assume alert == no results / an error.

is a fair assumption for clients to make, given what we document in the GQL API. Taking it at face value, we say:

> An alert message that should be displayed before any results.

In fact, this wording currently even implies that there _may_ be results (any) if there is an alert message. So the part about clients assuming alert -> no results seems unfair, and the part about assuming alert -> an error can remain intact (the client doesn't _have_ to look at whether there are results). It seems to me the alert -> no results is something we've abided by in the backend as an implicit convention, but not something we've asked clients to commit to. So our act of adding that example change would have the same bearing on clients (treat alerts as errors) as we currently do in the spec, and in the proposed spec, and I'm not suggesting a change to the GQL description nor the contract of clients-should-treat-alerts-as-errors.
